proposals/pattern-based-fixed.md

```diff
@@ -0,0 +1,135 @@
+# Pattern-based `fixed` statement
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Not Started
+
+## Summary
+[summary]: #summary
+
+Introduce a pattern that would allow types to participate in `fixed` statements. 
+
+## Motivation
+[motivation]: #motivation
+
+The language provides a mechanism for pinning managed data and obtain a native pointer to the underlying buffer.
+
+```C#
+fixed(byte* ptr = byteArray)
+{
+   // ptr is a native pointer to the first element of the array
+   // byteArray is protected from being moved/collected by the GC for the duration of this block 
+}
+
+```
+
+The set of types that can participate in `fixed` is hardcoded and limited to arrays and `System.String`. Hardcoding "special" types does not scale when new primitives such as `ImmutableArray<T>`, `Span<T>`, `Utf8String` are introduced. 
+
+In addition, the current solution for `System.String` relies on a fairly rigid API. The shape of the API implies that `System.String` is a contiguous object that embeds UTF16 encoded data at a fixed offset from the object header. Such approach has been found problematic in several proposals that could require changes to the underlying layout. 
+It would be desirable to be able to switch to something more flexible that decouples `System.String` object from its internal representation for the purpose of unmanaged interop. 
+
+## Detailed design
+[design]: #detailed-design
+
+## *Pattern* ##
+A viable pattern-based �fixed� need to:
+-	Provide the managed references to pin the instance and to initialize the pointer (preferably this is the same reference)
+-	Convey unambiguously the type of the unmanaged element   (i.e. �char� for �string�)
+-	Prescribe the behavior in "empty" case when there is nothing to refer to. 
+-	Should not push API authors toward design decisions that hurt the use of the type outside of `fixed`.
+
+
+I think the above could be satisfied by recognizing a specially named ref-returning member:
+ `ref [readonly] T DangerousGetPinnableReference()`.
+
+In order to be used by the `fixed` statement the following conditions must be met:
+
+1)	There is only one such member provided for a type.
+1)	Returns by `ref` or `ref readonly`. 
+(readonly is permitted so that authors of immutable/readonly types could implement the pattern)
+1)  T is an unmanaged type.
+(since `T*` becomes the pointer type)
+1)	Returns managed `nullptr` when there is no data to pin � probably the cheapest way to convey emptiness.
+(note that �� string returns a ref to �\0� since strings are zero-terminated)
+
+
+Alternatively for the `#3` we can allow the result in empty cases be undefined or implementation-specific. 
```

Comment:
There is difference in the strength of these conditions: 1) and 2) are enforced by the compiler; 3) is just a recommendation for the implementers of the `DangerousGetPinnableReference` API.

Reply:
Yes, it is just a recommendation, but I believe we have to recommend what to do, so that end users could have uniform expectations. 
(and the bugs about unexpected behavior would go to the type owners).